deps/ccommon/rust/ccommon_rs/src/buf.rs

```diff
@@ -0,0 +1,292 @@
+    pub fn rshift(&mut self) {
+        let size = self.read_size();
+
+        if size > 0 {
+            unsafe {
+                std::ptr::copy(
+                    self.buf.rpos,
+                    self.buf.rpos.offset(-(size as isize)),
+                    size
+                );
+            }
+        }
+
+        self.buf.rpos = unsafe { self.buf.end.offset(-(size as isize)) };
+        self.buf.wpos = self.buf.end;
+    }
+
+    /// Shift data in the buffer to the start of the buffer.
+    pub fn lshift(&mut self) {
+        let size = self.read_size();
+
+        if size > 0 {
+            unsafe {
+                std::ptr::copy(
+                    self.buf.rpos,
+                    self.buf.begin.as_mut_ptr(),
+                    size
+                );
+            }
+        }
+
+        self.buf.rpos = unsafe { self.buf.begin.as_mut_ptr() };
+        self.buf.wpos = self.buf.rpos.wrapping_add(size);
+    }
+
+    /// Get the currently valid part of the buffer as a slice
+    pub fn as_slice(&self) -> &[u8] {
+        use std::slice;
+
+        unsafe {
+            slice::from_raw_parts(
+                self.buf.rpos as *const u8,
+                self.read_size()
+            )
+        }
+    }
+
+    /// Get the currently valid part of the buffer as a slice
+    pub fn as_slice_mut(&mut self) -> &mut [u8] {
+        use std::slice;
+
+        unsafe {
+            slice::from_raw_parts_mut(
+                self.buf.rpos as *mut u8,
+                self.read_size()
+            )
+        }
+    }
+}
+
+/// An owned buffer.
+/// 
+/// In addition to all the operations supported by `Buf`
+/// an `OwnedBuf` also supports some resizing operations.
+#[repr(transparent)]
+pub struct OwnedBuf {
+    buf: *mut buf
+}
+
+impl OwnedBuf {
+    pub unsafe fn from_raw(buf: *mut buf) -> Self {
+        Self { buf }
+    }
+    pub fn into_raw(self) -> *mut buf {
+        let buf = self.buf;
+        // Don't want to run drop on the destructor
+        std::mem::forget(self);
+        buf
+    }
+
+    /// Double the size of the buffer.
+    pub fn double(&mut self) -> Result<(), crate::Error> {
+        use cc_binding::dbuf_double;
+
+        unsafe {
+            let status = dbuf_double(&mut self.buf as *mut _);
+
+            if status != 0 {
+                Err(status.into())
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Shrink the buffer to the max of the initial size or the content size.
+    pub fn shrink(&mut self) -> Result<(), crate::Error> {
+        use cc_binding::dbuf_shrink;
+
+        unsafe {
+            let status = dbuf_shrink(&mut self.buf as *mut _);
+
+            if status != 0 {
+                Err(status.into())
+            } else {
+                Ok(())
+            }
+        }
+    }
+
+    /// Resize the buffer to fit the required capacity.
+    /// 
+    /// # Panics
+    /// Panics if `cap` is greater than `u32::MAX`.
```

Comment:
Do we really want to take a `usize` here then? Might make more sense to take a `u32` size?

Reply:
Generally anything in rust that represents a size is done as a `usize`. If we use a `u32` here then I suspect most call sites will just do `some_usize as u32` which would silently give wrong results when `some_usize > u32::MAX`. That being said, we can change this if you think that `u32` is a better choice.

Reply:
Per discussion. OK to leave as usize, fix the assert to be `<=`

Reply:
For the scope of this project (i.e. we don't expect ccommon to be widely used by itself) it seems fine that we use `u32` to simplify the signature.